store/v2/multi/snapshot.go

```diff
@@ -0,0 +1,252 @@
+package root
+
+import (
+	"bufio"
+	"compress/zlib"
+	"fmt"
+	prefixdb "github.com/cosmos/cosmos-sdk/db/prefix"
+	"github.com/cosmos/cosmos-sdk/snapshots"
+	snapshottypes "github.com/cosmos/cosmos-sdk/snapshots/types"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	types "github.com/cosmos/cosmos-sdk/store/v2"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	protoio "github.com/gogo/protobuf/io"
+	"io"
+	"math"
+	"sort"
+	"strings"
+)
+
+// Snapshot implements snapshottypes.Snapshotter.
+func (rs *Store) Snapshot(height uint64, format uint32) (<-chan io.ReadCloser, error) {
+	if format != snapshottypes.CurrentFormat {
+		return nil, sdkerrors.Wrapf(snapshottypes.ErrUnknownFormat, "format %v", format)
+	}
+
+	if height == 0 {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "cannot snapshot height 0")
+	}
+	if height > uint64(rs.LastCommitID().Version) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrLogic, "cannot snapshot future height %v", height)
+	}
+	versions, err := rs.stateDB.Versions()
+	if err != nil {
+		return nil, sdkerrors.Wrapf(err, "error while getting the snapshot versions at height %v", height)
+	}
+	if !versions.Exists(height) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "cannot find snapshot at height %v", height)
+	}
+
+	// get the saved snapshot at height
+	vs, err := rs.getView(int64(height))
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, fmt.Sprintf("error while get the version at height %d", height))
+	}
+
+	// Spawn goroutine to generate snapshot chunks and pass their io.ReadClosers through a channel
+	ch := make(chan io.ReadCloser)
+	go func() {
+		// Set up a stream pipeline to serialize snapshot nodes:
+		// Export KV Item -> delimited Protobuf -> zlib -> buffer -> chunkWriter -> chan io.ReadCloser
+		chunkWriter := snapshots.NewChunkWriter(ch, snapshotChunkSize)
+		defer chunkWriter.Close()
+		bufWriter := bufio.NewWriterSize(chunkWriter, snapshotBufferSize)
+		defer func() {
+			if err := bufWriter.Flush(); err != nil {
+				chunkWriter.CloseWithError(err)
+			}
+		}()
+		// zlib compression levels:  https://www.euccas.me/zlib/#zlib_compression_levels
+		zWriter, err := zlib.NewWriterLevel(bufWriter, 7)
```

Comment:
Can we more explicitly type what level of compression "7" specifies by defining it as properly named constant variable for it? The comment above is good, but it would be nice to have an idea what "7" means without reading through the above docs.

Reply:
Ok